cmd/lncli/commands.go

```diff
@@ -672,51 +679,110 @@ func closeChannel(ctx *cli.Context) error {
 		req.ChannelPoint.OutputIndex = 0
 	}
 
-	stream, err := client.CloseChannel(ctxb, req)
+	closingTxid, err := closeChannelRequest(ctx, req)
 	if err != nil {
 		return err
 	}
 
+	printJSON(struct {
+		ClosingTxid string `json:"closing_txid"`
+	}{
+		ClosingTxid: closingTxid,
+	})
+
+	return nil
+}
+
+func closeChannelRequest(ctx *cli.Context, req *lnrpc.CloseChannelRequest) (string, error) {
+	client, cleanUp := getClient(ctx)
+	defer cleanUp()
+
+	stream, err := client.CloseChannel(context.Background(), req)
+	if err != nil {
+		return "", err
+	}
+
+	var closingTxid string
+
 	for {
 		resp, err := stream.Recv()
 		if err == io.EOF {
-			return nil
+			return closingTxid, nil
 		} else if err != nil {
-			return err
+			return "", err
 		}
 
 		switch update := resp.Update.(type) {
 		case *lnrpc.CloseStatusUpdate_ClosePending:
 			closingHash := update.ClosePending.Txid
 			txid, err := chainhash.NewHash(closingHash)
 			if err != nil {
-				return err
+				return "", err
 			}
 
-			printJSON(struct {
-				ClosingTXID string `json:"closing_txid"`
-			}{
-				ClosingTXID: txid.String(),
-			})
+			closingTxid = txid.String()
 
 			if !ctx.Bool("block") {
-				return nil
+				return closingTxid, nil
 			}
-
 		case *lnrpc.CloseStatusUpdate_ChanClose:
 			closingHash := update.ChanClose.ClosingTxid
 			txid, err := chainhash.NewHash(closingHash)
 			if err != nil {
-				return err
+				return "", err
 			}
 
-			printJSON(struct {
-				ClosingTXID string `json:"closing_txid"`
-			}{
-				ClosingTXID: txid.String(),
-			})
+			closingTxid = txid.String()
+		}
+	}
+}
+
+func closeAllChannels(ctx *cli.Context) error {
+	ctxb := context.Background()
+	client, cleanUp := getClient(ctx)
+	defer cleanUp()
+
+	listReq := &lnrpc.ListChannelsRequest{}
+	openChannels, err := client.ListChannels(ctxb, listReq)
+	if err != nil {
+		return fmt.Errorf("unable to fetch open channels: %v", err)
+	}
+
+	txids := make([]string, len(openChannels.Channels))
+	for i, channel := range openChannels.Channels {
+		s := strings.Split(channel.ChannelPoint, ":")
+		txid, err := chainhash.NewHashFromStr(s[0])
+		if err != nil {
+			continue
+		}
+		index, err := strconv.ParseUint(s[1], 10, 32)
+		if err != nil {
+			continue
+		}
+
+		req := &lnrpc.CloseChannelRequest{
+			ChannelPoint: &lnrpc.ChannelPoint{
+				FundingTxid: txid[:],
+				OutputIndex: uint32(index),
+			},
+			Force: !channel.Active,
+		}
+
+		closingTxid, err := closeChannelRequest(ctx, req)
+		if err != nil {
+			continue
```

Comment:
It would probably be good to do some error reporting if a channel fails to close (especially since in the existing channel closure case, if a channel fails to close, there is an error reported to the user). I'm not sure what the most idiomatic way to do that for this project is, but I'd I see two main ways to do it (and it is possible that both should be done).

1. Print an error message upon each closure here, before the "continue". I'm not sure if printing error messages as you go is idiomatic though.
2. Return an error message at the end of this function if the number of channels that were closed was fewer than the number of channels that were requested to be closed. This might be more idiomatic, but has the down side of the user not knowing why individual channels didn't close.

You'll probably need feedback from a regular contributor on how best to handle this, though I suspect that at least some error reporting is needed.

Reply:
Yeah, some error reporting would definitely be helpful. I wanted to submit this first just for some feedback. I think something like the following response would be helpful:

```
{
    "completed": [
        {
            "remote_pubkey": "...",
            "channel_point": "...",
            "closing_txid": "...",
        },
        {
            "remote_pubkey": "...",
            "channel_point": "...",
            "closing_txid": "...",
        },
    ],
    "failed": [
        {
            "remote_pubkey": "...",
            "channel_point": "...",
            "error": "...",
        },
    ],
}
```

Reply:
Indeed. For example, on mainnet we may want to prompt the user if they would like to _really_ force close, showing the number of blocks they'd need to wait until the funds can be swept entirely.